sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobServiceSasSignatureValues.java

```diff
@@ -474,60 +478,111 @@ public BlobServiceSasQueryParameters generateSasQueryParameters(SharedKeyCredent
      * encoded string, or the UTF-8 charset isn't supported.
-     * @throws NullPointerException if {@code delegationKey} is null. Or if any of {@code version},
-     * {@code canonicalName}, {@code resource}, {@code expiryTime} or {@code permissions} are null.
+     * @throws IllegalArgumentException if {@link #getPermissions()} contains an invalid character for the SAS resource.
+     * @throws NullPointerException if {@code delegationKey} or {@code account} is null.
+     * @see <a href="https://docs.microsoft.com/rest/api/storageservices/create-user-delegation-sas">Create a user
+     * delegation SAS</a>
      */
-    public BlobServiceSasQueryParameters generateSasQueryParameters(UserDelegationKey delegationKey) {
+    public BlobServiceSasQueryParameters generateSasQueryParameters(UserDelegationKey delegationKey, String accountName) {
         Utility.assertNotNull("delegationKey", delegationKey);
-        assertGenerateOK(true);
+        Utility.assertNotNull("accountName", accountName);
+
+
+        ensureState();
 
         // Signature is generated on the un-url-encoded values.
-        String signature = Utility.computeHMac256(delegationKey.getValue(), stringToSign(delegationKey));
+        final String canonicalName = getCanonicalName(accountName);
+        String signature = Utility.computeHMac256(delegationKey.getValue(), stringToSign(delegationKey, canonicalName));
 
         return new BlobServiceSasQueryParameters(this.version, this.protocol, this.startTime, this.expiryTime,
             this.sasIpRange, null /* identifier */, this.resource, this.permissions, signature, this.cacheControl,
             this.contentDisposition, this.contentEncoding, this.contentLanguage, this.contentType, delegationKey);
     }
 
     /**
-     * Common assertions for generateSASQueryParameters overloads.
+     * Ensures that the builder's properties are in a consistent state.
+
+     * 1. If there is no version, use latest.
+     * 2. Resource name is chosen by:
+     *    a. If "BlobName" is _not_ set, it is a container resource.
+     *    b. Otherwise, if "SnapshotId" is set, it is a blob snapshot resource.
+     *    c. Otherwise, it is a blob resource.
+     * 3. Reparse permissions depending on what the resource is. If it is an unrecognised resource, do nothing.
+     *
+     * Taken from:
+     * https://github.com/Azure/azure-storage-blob-go/blob/master/azblob/sas_service.go#L33
+     * https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/storage/Azure.Storage.Blobs/src/Sas/BlobSasBuilder.cs
      */
-    private void assertGenerateOK(boolean usingUserDelegation) {
-        Utility.assertNotNull("version", this.version);
-        Utility.assertNotNull("canonicalName", this.canonicalName);
-        Utility.assertNotNull("resource", this.resource);
+    private void ensureState() {
+        if (version == null) {
+            version = Constants.HeaderConstants.TARGET_STORAGE_VERSION;
```

Comment:
Should this be changed based on the service version work that @samvaity has been doing (which is now in the repo)?